presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeometryOperators.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.plugin.geospatial;
+
+import com.facebook.presto.spi.function.IsNull;
+import com.facebook.presto.spi.function.ScalarOperator;
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.spi.type.StandardTypes;
+import io.airlift.slice.Slice;
+
+import static com.facebook.presto.geospatial.serde.GeometrySerde.deserialize;
+import static com.facebook.presto.plugin.geospatial.GeometryType.GEOMETRY_TYPE_NAME;
+import static com.facebook.presto.spi.function.OperatorType.EQUAL;
+import static com.facebook.presto.spi.function.OperatorType.HASH_CODE;
+import static com.facebook.presto.spi.function.OperatorType.IS_DISTINCT_FROM;
+import static com.facebook.presto.spi.function.OperatorType.NOT_EQUAL;
+
+public final class GeometryOperators
+{
+    private GeometryOperators() {}
+
+    @ScalarOperator(EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    @SqlNullable
+    public static Boolean equal(@SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlType(GEOMETRY_TYPE_NAME) Slice right)
+    {
+        if (left == null) {
+            return right == null;
+        }
+
+        if (right == null) {
+            return false;
+        }
+
+        return deserialize(left).Equals(deserialize(right));
+    }
+
+    @ScalarOperator(NOT_EQUAL)
+    @SqlType(StandardTypes.BOOLEAN)
+    @SqlNullable
+    public static Boolean notEqual(@SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlType(GEOMETRY_TYPE_NAME) Slice right)
+    {
+        return !equal(left, right);
+    }
+
+    @ScalarOperator(IS_DISTINCT_FROM)
+    @SqlType(StandardTypes.BOOLEAN)
+    public static boolean isDistinctFrom(
+            @SqlType(GEOMETRY_TYPE_NAME) Slice left,
+            @IsNull boolean leftNull,
+            @SqlType(GEOMETRY_TYPE_NAME) Slice right,
+            @IsNull boolean rightNull)
+    {
+        if (leftNull != rightNull) {
+            return true;
+        }
+        if (leftNull) {
+            return false;
+        }
+        return notEqual(left, right);
+    }
+
+    @ScalarOperator(HASH_CODE)
+    @SqlType(StandardTypes.BIGINT)
+    public static long hashCode(@SqlType(GEOMETRY_TYPE_NAME) Slice slice)
+    {
+        if (slice == null) {
+            return 0;
+        }
+        return deserialize(slice).hashCode();
```

Comment:
This returns a 32bit hash code, so it may potentially cause skews. it would be nice to compute a 64bit hash from the Geomentry object, but i understand that it is very nontrivial. Do you foresee any large aggregations / joins on Geometry keys?

Also, when casted to 64bit long the first 32 bits will be zeros (for a positive value at least). I would suggest taking a xxhash64 from the 32bit hashcode returned by the `Geometry#hashCode()`. It won't give you more entropy, but at least the existing entropy will be evenly distributed across all of the 64 bits.

Reply:
I don't expect aggregations or joins on geometries other than points, but for points I do think there could be a few. Would it be helpful to specialize this function for points or use `XxHash64.hash(slice)` on the Slice itself (without even deserializing to geometry)?

Reply:
If you are sure that the hashcode contract won't be broken (if A=B then hashcode(A) = hashcode(B)), in other words if Point is always serialized the same way - i think it would be reasonable. Otherwise you can always deserialize the Point (AFAIK Point deserialization is cheap).